logs/parse.go

```diff
@@ -1,519 +1,249 @@
-			parts := LogPrefixHeroku1Regexp.FindStringSubmatch(line)
-			if len(parts) == 0 {
-				return
-			}
-			// skip %e
-			levelPart = parts[2]
-			contentPart = parts[3]
-		case LogPrefixHeroku2:
-			parts := LogPrefixHeroku2Regexp.FindStringSubmatch(line)
-			if len(parts) == 0 {
-				return
-			}
-			// skip %e
-			timePart = parts[2]
-			pidPart = parts[3]
-			// skip %s
-			// skip %c
-			// skip %v
-			// skip %x
-			logLineNumberPart = parts[8]
-			dbPart = parts[9]
-			// skip %r
-			userPart = parts[11]
-			appPart = parts[12]
-			levelPart = parts[13]
-			contentPart = parts[14]
-		default:
-			// Some callers use the content of unparsed lines to stitch multi-line logs together
-			logLine.Content = line + lineExtra
-			return
-		}
+	if parserInSync {
+		return
 	}
 
-	if timePart != "" {
-		occurredAt := GetOccurredAt(timePart, tz, rsyslog)
-		if occurredAt.IsZero() {
-			return
-		}
-		logLine.OccurredAt = occurredAt
-	}
+	server.LogParseMutex.Lock()
+	defer server.LogParseMutex.Unlock()
 
-	if userPart != "[unknown]" {
-		logLine.Username = userPart
-	}
-	if dbPart != "[unknown]" {
-		logLine.Database = dbPart
-	}
-	if appPart != "[unknown]" {
-		logLine.Application = appPart
+	server.LogParser = NewLogParser(prefix, tz, isSyslog, useLegacyFallback)
+}
+
+func (lp *LogParser) ValidatePrefix() error {
+	dbInPrefix, err := regexp.MatchString("(?:^|[^%])%d", lp.prefix)
+	if err != nil {
+		return fmt.Errorf("could not check: %s", err)
 	}
-	if logLineNumberPart != "" {
-		logLineNumber, _ := strconv.ParseInt(logLineNumberPart, 10, 32)
-		logLine.LogLineNumber = int32(logLineNumber)
+	userInPrefix, err := regexp.MatchString("(?:^|[^%])%u", lp.prefix)
+	if err != nil {
+		return fmt.Errorf("could not check: %s", err)
 	}
-
-	backendPid, _ := strconv.ParseInt(pidPart, 10, 32)
-	logLine.BackendPid = int32(backendPid)
-	logLine.Content = contentPart + lineExtra
-
-	// This is actually a continuation of a previous line
-	if levelPart == "" {
-		return
+	if !dbInPrefix && !userInPrefix {
+		return errors.New("database (%d) and user (%u) not found: pganalyze will not be able to correctly classify some log lines")
+	} else if !dbInPrefix {
+		return errors.New("database (%d) not found: pganalyze will not be able to correctly classify some log lines")
+	} else if !userInPrefix {
+		return errors.New("user (%u) not found: pganalyze will not be able to correctly classify some log lines")
+	} else {
+		return nil
 	}
+}
 
-	logLine.LogLevel = pganalyze_collector.LogLineInformation_LogLevel(pganalyze_collector.LogLineInformation_LogLevel_value[levelPart])
-	ok = true
-
-	return
+func (lp *LogParser) Matches(prefix string, tz *time.Location, isSyslog bool) bool {
+	return lp.prefix == prefix && tz.String() == lp.tz.String() && lp.isSyslog == isSyslog
 }
 
-func GetOccurredAt(timePart string, tz *time.Location, rsyslog bool) time.Time {
-	if tz != nil && !rsyslog {
+func (lp *LogParser) GetOccurredAt(timePart string) time.Time {
```

Comment:
memo: this is pretty much the same logic as previous one (well, I guess it's obvious from this diff)